[vcl] Better debug

gquintard opened the pull request:
Detecting if the response is a hit usingthe `x-varnish` header is flaky
and notably doesn't work in a multi-tier setup.

On top of this, HIT/MISS is a false dichotomy (https://info.varnish-software.com/blog/using-obj-hits)
and while I'd like something more detailed like
https://docs.varnish-software.com/tutorials/hit-miss-logging/ (tech
version of the blog), I recognize that it's "a lot" of VCL, so this
commit only differentiates between hits, misses and passes, the latter
being pretty useful to identify while debugging.

note: all versions have been changed for the sake of consistency
but both the 4.x and 5.x series have been EOL'd a (long) while ago and users
should be encouraged to upgraded as soon as possible.

*edit:* I've added a test that you can run with `varnishtest dev/tests/varnish/*.vtc` since the current `php` suite doesn't correctly test the committed VCL. It would probably be smart to run that command as part of the test suite, but I'll leave that to more knowledgeable people than myself.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#37912: [vcl] Better debug

gquintard commented:
@magento run all tests

lbajsarowicz commented:
@magento run Unit Tests

lbajsarowicz reviewed (dismissed):
:heavy_check_mark: Thank you for your expertise!

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7766](https://jira.corp.magento.com/browse/ENGCOM-7766) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

lbajsarowicz commented:
@magento run Integration Tests, Functional Tests CE

gquintard commented:
@lbajsarowicz, apologies for my ignorance, is there anything I should do here?

lbajsarowicz commented:
@gquintard No. Now we wait for the process of the merge to be performed.

lbajsarowicz commented:
**Automated Tests** are not applicable, as the change is related to infrastructure configuration.

engcom-Bravo commented:
Dev experience is required for testing of this PR. Please note that Manual testing has not been performed.

ihor-sviziev commented:
@engcom-Bravo  seems like you missed to assign the PR. I just did it for you

engcom-Charlie reviewed (changes requested):
@gquintard during testing on varnish 4.1.11 I faced an issue by following steps
1. Go to `Stores` > `Settings` > `Configuration` > `Advanced` > `System` > `Full Page Cache`
2. Expand `Varnish Configuration`
3. Fill needed fields
4. Export VCL config
5. Copy it to `/etc/varnish/` as `default.vcl`
6. Restart varnish

**AR**: 
`Job for varnish.service failed because the control process exited with error code. See "systemctl status varnish.service" and "journalctl -xe" for details.`
**Error in log**:
`Failed to start Varnish Cache, a high-performance HTTP accelerator.`

ihor-sviziev commented:
@engcom-Charlie basicwlly varnish 4 already reached end of life https://info.varnish-software.com/blog/options-at-varnish-cache-4-end-of-support . I believe if everything works fine on 5 and 6 - let’s revert changes in varnish4 config

gquintard commented:
Indeed, `obj.uncacheable` was introduced later, I missed that, sorry.

@ihor-sviziev , 5.X is also EOL, if you want to just focus on 6.X: https://varnish-cache.org/releases/index.html

ihor-sviziev commented:
@gquintard could you just revert changes for varnish4 and varnish5 than?

gquintard commented on app/code/Magento/PageCache/etc/varnish4.vcl:
```diff
@@ -196,7 +196,9 @@ sub vcl_backend_response {
 sub vcl_deliver {
     if (resp.http.X-Magento-Debug) {
-        if (resp.http.x-varnish ~ " ") {
+        if (obj.uncacheable) {
+            set resp.http.X-Magento-Cache-Debug = "UNCACHEABLE";
+        } else if (obj.hits) {
```
```suggestion
        if (resp.http.x-varnish ~ " ") {
```

gquintard commented on app/code/Magento/PageCache/etc/varnish5.vcl:
```diff
@@ -197,7 +197,9 @@ sub vcl_backend_response {
 sub vcl_deliver {
     if (resp.http.X-Magento-Debug) {
-        if (resp.http.x-varnish ~ " ") {
+        if (obj.uncacheable) {
+            set resp.http.X-Magento-Cache-Debug = "UNCACHEABLE";
+        } else if (obj.hits) {
```
```suggestion
        if (resp.http.x-varnish ~ " ") {
```

gquintard commented:
done

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (approved):
✔ Approved.

Failing tests looks not related to changes form this PR.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7766](https://jira.corp.magento.com/browse/ENGCOM-7766) has been created to process this Pull Request

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests EE

gquintard commented:
functional tests EE are failing, but I don't see anyway for them to fail the way they do. False negative?

ihor-sviziev commented:
@magento run Functional Tests EE

ihor-sviziev commented:
@magento run Functional Tests EE

gquintard commented:
bump?

gquintard commented:
hi, we are almost at the one-year anniversary of that PR, is there something I can do to make it happen?

ihor-sviziev commented:
@gquintard, unfortunately, this PR has low priority and will be processed after all high priority ones

mrtuvn commented:
@gquintard can you resolve conflicts

gquintard commented:
@mrtuvn , done

mrtuvn commented:
@magento run all tests. Rerun checks

engcom-Lima commented:
@magento create issue

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

engcom-Hotel commented:
Hello @gquintard,

Thanks for the contribution!

We have tried to reproduce the issue with the 2.4-develop branch as well as PR changes. But the difference is not visible in both of them.

We request you to please provide the `Manual Testing Scenarios` to test this PR.

Thanks

engcom-Hotel commented:
@magento run Functional Tests B2B, WebAPI Tests

engcom-Hotel commented:
@magento run Functional Tests B2B

engcom-Hotel commented:
Hello @gquintard,

Have you had a chance to check this comment https://github.com/magento/magento2/pull/28944#issuecomment-1695651359?

Actually, the issue is not reproducible for us, can you please provide us with the manual testing scenario? Meanwhile, we are moving this issue `On Hold`.

Thanks

engcom-Hotel commented:
@magento run all tests

gquintard commented:
I've added a test, but I'm not sure that what's failing here is related to that, I'll poke at it later this week, but if anyone has pointers, I'll take them

engcom-Hotel commented:
Hello @gquintard,

Thanks for adding the automated tests with this PR. We request you to check the build failures, it seems the tests are failing due to code changes.

Also as per our previous comment, we need to have some `Manual testing scenarios` to test this PR. Please provide those as well.

Thanks

engcom-Hotel commented:
Hello @gquintard,

As I can see you have added some automated tests in this PR, due to which the builds are failing.

> I've added a test, but I'm not sure that what's failing here is related to that, I'll poke at it later this week, but if anyone has pointers, I'll take them

So I suggested you check the integration test failure and Functional Test failure CE on the below link:

Integration test failure: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/142665397fee33cf8fc698645b34fa94/Integration/console-error-logs.html

Functional Test CE: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/66edf9bd8343e16738fe5a0a6438a8c6/Functional/allure-report-ce/index.html#categories

Thanks

gquintard commented:
@magento run all tests

gquintard commented:
I was unable to understand the link between my VCL changes, the failing test and my added tests (which are different), so I reverted my commits, and the CI is still failing, so I doubt it can be trusted.

engcom-Hotel commented:
Hello @gquintard,

Actually, in this comment https://github.com/magento/magento2/pull/28944#issuecomment-1715346316, you can see an integration test failure as below:

```
1) Magento\GraphQlCache\Controller\Cms\BlockCacheTest::testCmsBlocksRequestHasCorrectTags
InvalidArgumentException: Unable to unserialize value. Error: Syntax error

/var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php:45
/var/www/html/app/code/Magento/GraphQl/Controller/GraphQl.php:237
/var/www/html/app/code/Magento/GraphQl/Controller/GraphQl.php:186
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:58
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:138
/var/www/html/app/code/Magento/PageCache/Model/App/FrontController/BuiltinPlugin.php:76
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:135
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:153
/var/www/html/dev/tests/integration/tmp/sandbox--61c9aa5837ee9e5acb35117401d1dfba293827a5eba4de85822ec7f3c2ea2f97/generated/code/Magento/GraphQl/Controller/GraphQl/Interceptor.php:23
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/AbstractGraphqlCacheTest.php:216
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/Cms/BlockCacheTest.php:118
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/AbstractGraphqlCacheTest.php:86

ERRORS!
Tests: 29, Assertions: 77, Errors: 1.
```

I was talking about this, You can see after reverting all the changes the same error is not occurring in the below integration test failures:

https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/45585c83690cf31ea2bb23a421e40fb9/Integration/console-error-logs.html

You can ignore Functional Test failures for now it seems flaky to me.

Thanks

engcom-Hotel commented:
Hello @gquintard,

Actually, in this comment https://github.com/magento/magento2/pull/28944#issuecomment-1715346316 we are talking about the below failure:

```
1) Magento\GraphQlCache\Controller\Cms\BlockCacheTest::testCmsBlocksRequestHasCorrectTags
InvalidArgumentException: Unable to unserialize value. Error: Syntax error

/var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php:45
/var/www/html/app/code/Magento/GraphQl/Controller/GraphQl.php:237
/var/www/html/app/code/Magento/GraphQl/Controller/GraphQl.php:186
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:58
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:138
/var/www/html/app/code/Magento/PageCache/Model/App/FrontController/BuiltinPlugin.php:76
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:135
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:153
/var/www/html/dev/tests/integration/tmp/sandbox--61c9aa5837ee9e5acb35117401d1dfba293827a5eba4de85822ec7f3c2ea2f97/generated/code/Magento/GraphQl/Controller/GraphQl/Interceptor.php:23
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/AbstractGraphqlCacheTest.php:216
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/Cms/BlockCacheTest.php:118
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/AbstractGraphqlCacheTest.php:86

ERRORS!
Tests: 29, Assertions: 77, Errors: 1.
```
The changes have been occurring due to the PR changes, and you can see in the below integration test failure the same error is not occurring:

https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/45585c83690cf31ea2bb23a421e40fb9/Integration/console-error-logs.html

You can ignore functional test failures, for now, it seems flaky to me.

Thanks

gquintard commented:
@magento run all tests

gquintard commented:
@magento run all tests

gquintard commented:
@magento run all tests

engcom-Charlie commented:
Hi @gquintard,

Thank you for your contribution!

Did you get a chance to look into test failures mentioned [here](https://github.com/magento/magento2/pull/28944#issuecomment-1715346316).

Thank you!

gquintard commented:
not yet, I will get to it eventually though
-- 
Guillaume Quintard


On Wed, Oct 4, 2023 at 2:47 AM Indrani ***@***.***> wrote:

> Hi @gquintard <https://github.com/gquintard>,
>
> Thank you for your contribution!
>
> Did you get a chance to look into test failures mentioned here
> <https://github.com/magento/magento2/pull/28944#issuecomment-1715346316>.
>
> Thank you!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento2/pull/28944#issuecomment-1746521767>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA42AKJEASEPNCHPUELYUH3X5UWDFAVCNFSM4OMVWQS2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCNZUGY2TEMJXGY3Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gquintard commented:
@magento run all tests

gquintard commented:
by the way, is there an easy way to run those test locally, and possibly with a more narrow scope? Having to go though CI every time is fairly time consuming

engcom-Hotel commented:
Hello @gquintard,

I think you are looking for integration tests running locally and you need to run any specific test. For the same, you can go through with the below DevDocs:

https://developer.adobe.com/commerce/testing/guide/integration/#run-a-single-test-class

Let us know in case you need more assistance.

Thanks

gquintard commented:
thanks @engcom-Hotel , to be fair, setting up testing appears to be a larger job than I care for, specially for such a minor patch, so I'll just rely on automated testing

gquintard commented:
@engcom-Hotel, could you have a look at the diff and see if something obvious stands out?

I'm getting this error:
```

1) Magento\GraphQlCache\Controller\Cms\BlockCacheTest::testCmsBlocksRequestHasCorrectTags
InvalidArgumentException: Unable to unserialize value. Error: Syntax error

/var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php:45
/var/www/html/app/code/Magento/GraphQl/Controller/GraphQl.php:237
/var/www/html/app/code/Magento/GraphQl/Controller/GraphQl.php:186
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:58
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:138
/var/www/html/app/code/Magento/PageCache/Model/App/FrontController/BuiltinPlugin.php:76
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:135
/var/www/html/lib/internal/Magento/Framework/Interception/Interceptor.php:153
/var/www/html/dev/tests/integration/tmp/sandbox--61c9aa5837ee9e5acb35117401d1dfba293827a5eba4de85822ec7f3c2ea2f97/generated/code/Magento/GraphQl/Controller/GraphQl/Interceptor.php:23
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/AbstractGraphqlCacheTest.php:217
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/Cms/BlockCacheTest.php:118
/var/www/html/dev/tests/integration/testsuite/Magento/GraphQlCache/Controller/AbstractGraphqlCacheTest.php:86
```

As `dispatchGraphQlGETRequest` is essentially a copy of `dispatchGraphQlGETRequest` and as the error message isn't telling much about the serialization, I'm scracting my head here

hostep commented:
If I have to _guess_, I think that the content needs to be valid json syntax, so I would try replacing: `$request->setContent('');` with `$request->setContent('{}');` or `$request->setContent('""');`
Maybe that will work?

engcom-Hotel commented:
@magento run all tests

gquintard commented:
@magento run all tests

gquintard commented:
@magento run all tests

engcom-Hotel commented:
@magento run Integration Tests, Functional Tests B2B

gquintard commented:
@engcom-Hotel , is there *any* chance that the test environment isn't using the committed VCL?
I'm looking at https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/5852bb826f8f6c55f28339b9670a2fe2/Integration/console-error-logs.html and the actual value of the `X-Magento-Cache-Debug` is `MISS`, which shouldn't be possible, since now the VCL only sets that header to `MISS4`/`MISS5`/`MISS6`

engcom-Hotel commented:
Hello @gquintard,

It should use the committed VCL file only, let me check at our end. We will get back to you soon.

Thanks

gquintard commented:
thanks, does that mean that there's no VCL testing then?

engcom-Hotel commented:
Thats right @gquintard

engcom-Hotel commented:
Hello @gquintard,

Are you still working on the failed test for this PR?

Thanks

gquintard commented:
I am, but I need to come up with a new series of tests since the current framework just doesn't actually test the vcl

engcom-Hotel commented:
Thanks @gquintard for the reply!

As you are still working on this PR, we are marking this issue as a draft. 

Thanks

gquintard committed:
[vcl] Better debug

Detecting if the response is a hit usingthe x-varnish header is flaky
and notably doesn't work in a multi-tier setup.

On top of this, HIT/MISS is a false dichotomy (https://info.varnish-software.com/blog/using-obj-hits)
and while I'd like something more detailed like
https://docs.varnish-software.com/tutorials/hit-miss-logging/ (tech
version of the blog), I recognize that it's "a lot" of VCL, so this
commit only differentiates between hits, misses and passes, the latter
being pretty useful to identify while debugging.

gquintard commented:
@magento run all tests

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on dev/tests/varnish/README.md:
```diff
@@ -0,0 +1,11 @@
+
+Documentation:
+- varnishtest itself: https://varnish-cache.org/docs/trunk/reference/varnishtest.html
+- VTC sytntax: https://varnish-cache.org/docs/trunk/reference/vtc.html
```
Please correct the `syntax` typo

gquintard replied:
thanks, done

engcom-Hotel commented on dev/tests/varnish/debug.vtc:
We prefer automated tests so that they can be run automatically. It seems this test will not run automatically, we need to run this test manually. Please confirm.

gquintard replied:
yes, this was the point of my message. For now its manual, but if you know where and how to hook it up to the automated testing it would be best to do that.

engcom-Hotel reviewed (changes requested):
Hello @gquintard,

Thanks for the collaboration!

Please find my review comment below.

Thanks

engcom-Hotel commented:
Hello @gquintard,

Started discussion on this PR internally. Meanwhile, we are moving this PR on hold. We will update you soon on this.

Thanks

gquintard commented:
Thanks! I have a handful of other VCL cleanup/fixes lined up, and I feel that using `varnishtest` makes changes a lot more robust, so I would be inclined to opening more PRs with the same kind of test cases. The question is: should I wait for this PR to actually land, or should I go ahead and start working on the others?

engcom-Hotel commented:
Hello @gquintard,

What I suggest here is, if your PR's tests contain `varnishtest` then please wait for this PR otherwise you can continue with the other PRs.

Thanks

engcom-Hotel commented:
:x: QA not Passed

**Preconditions:**
Detecting if the response is a hit using the x-varnish header is flaky and notably doesn't work in a multi-tier setup.

**Manual testing scenario:**

Running the below command to check test is passed or failed:
`varnishtest dev/tests/varnish/debug.vtc`

**Actual Result:** :heavy_check_mark:
The above command should return as Passed.

**After:** : heavy_multiplication_x:  

Attaching the log of varnishtest. We have replaced the local file path with `/DummyPath/`

gquintard committed:
[varnish] fix debug.vtc on mac

gquintard commented:
@engcom-Hotel , thank you for your patience. I was able to get access to a mac thanks to the awesome @briiians, and it looks like on it, `sed FILE -e EXPRESSION` doesn't work (`FILE` is interpreted as an expression).

So, normally, this should be fixed, please have another look

engcom-Hotel committed:
Merge branch '2.4-develop' into vcl_debug

andrewbess commented:
@magento run all tests

andrewbess commented:
Hello @gquintard 
Thank you for your contribution
Could you please change commit messages to more relevant?
I think the message: "**_[varnish] fix debug.vtc on mac_**" is not clear and doesn't describe your fixes.
Thank you in advance

gquintard commented:
hi @andrewbess, sure. Can I do that once we are just about to merge? I'll rebase an squash at that time.

andrewbess commented:
> hi @andrewbess, sure. Can I do that once we are just about to merge? I'll rebase an squash at that time.

Yes, of course
Thank you

andrewbess reviewed (approved):
The fixes look nice for me
Approved

andrewbess commented:
@magento run Functional Tests EE, Functional Tests B2B

engcom-Hotel commented:
@magento run Functional Tests B2B

engcom-Hotel commented:
:heavy_check_mark: QA Passed

**Preconditions:**
This PR differentiates between hits, misses, and passes, the latter being pretty useful to identify while debugging.

**Manual testing scenario:**
Run the below `varnishtest` command to check whether it is working fine or not:

`varnishtest dev/tests/varnish/debug.vtc`

**Actual Result:** :heavy_check_mark:

The above test should show `passed` as below:
![image](https://github.com/magento/magento2/assets/51681618/bf2aeb72-530d-4504-99a5-c30c3a892529)


**Tested all the manual scenarios, no impact on regression testing.**

engcom-Hotel commented:
@magento run Functional Tests B2B

engcom-Hotel commented:
@magento run Functional Tests B2B

engcom-Hotel commented:
The failed test seems flaky to me. Please refer to the below screenshots:

[_**Failed Test 1**_](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/db6c6d11665ec168ef44cf7cb4be3518/Functional/allure-report-b2b/index.html#categories/bcaf5e97f075ebb22dc7e494f9bb82ca/c6523862f80b7ef1/)
![image](https://github.com/magento/magento2/assets/51681618/bfa798ef-62e5-42be-b5e1-e93fe1375c79)

[_**Failed Test 2**_](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28944/985595758dff02d386b3dfdbfe9a973b/Functional/allure-report-b2b/index.html#categories)
![image](https://github.com/magento/magento2/assets/51681618/9f88f554-b8cc-4b1a-a127-8e15b050a9ee)

gquintard commented:
We've established earlier that the VCL code changes aren't taken into account by automated tests (the test used the system-wide VCL which doesn't match the in-repository code). I do believe it's an issue in itself, but I'm fairly confident the flaky tests aren't caused by this change.
